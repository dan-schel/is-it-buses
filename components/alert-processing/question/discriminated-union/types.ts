import { ObjectField } from "@/components/alert-processing/question/object/ObjectField";
import { ObjectValue } from "@/components/alert-processing/question/object/types";

export type AnyDiscriminatedUnionConfig = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: ObjectField<any>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ExtractObjectConfig<Field extends ObjectField<any>> =
  Field extends ObjectField<infer Config> ? Config : never;

// TODO: Maybe this can be simplified, but it seems to work!
export type DiscriminatedUnionValue<
  Discriminator extends string,
  Config extends AnyDiscriminatedUnionConfig,
> = {
  [K in keyof Config]: { [D in Discriminator]: K } & ObjectValue<
    ExtractObjectConfig<Config[K]>
  >;
}[keyof Config];

export type SemiRawDiscriminatedUnionValue<
  Config extends AnyDiscriminatedUnionConfig,
> = {
  [K in keyof Config]: { type: K } & {
    value: ObjectValue<ExtractObjectConfig<Config[K]>> | null;
  };
}[keyof Config];

export type RawDiscriminatedUnionValue<
  Config extends AnyDiscriminatedUnionConfig,
> = SemiRawDiscriminatedUnionValue<Config> | { type: null };

export type TypesWithinUnion<Config extends AnyDiscriminatedUnionConfig> =
  keyof Config & string;
