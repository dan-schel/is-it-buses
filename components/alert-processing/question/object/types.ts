/* eslint-disable @typescript-eslint/no-explicit-any */

import { Field } from "@/components/alert-processing/question/lib/field";
import { Maybe } from "@/shared/types/maybe";

export type AnyField = Field<any, any>;

export type AnyConfigType = { [key: string]: AnyField };

type FieldValue<T extends Field<any, any>> =
  T extends Field<infer V, any> ? V : never;

export type ObjectValue<C extends AnyConfigType> = {
  [K in keyof C]: FieldValue<C[K]>;
};

export type RawObjectValue<C extends AnyConfigType> = {
  [K in keyof C]: Maybe<FieldValue<C[K]>>;
};

export type ObjectValidateFunction<Config extends AnyConfigType> = (
  input: ObjectValue<Config>,
) => { error: string; questionsToInvalidate: (keyof Config)[] } | null;
