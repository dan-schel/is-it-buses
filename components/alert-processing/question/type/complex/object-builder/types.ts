/* eslint-disable @typescript-eslint/no-explicit-any */

import { QuestionProps } from "@/components/alert-processing/question/lib/use-question";
import { DateQuestionAdditionalProps } from "@/components/alert-processing/question/type/DateQuestion";
import { StringQuestionAdditionalProps } from "@/components/alert-processing/question/type/StringQuestion";
import { Maybe } from "@/shared/types/maybe";
import React from "react";

export type Field<T, Extra> = {
  type: React.FC<QuestionProps<T, Extra>>;
  props: Extra;
};

export type AnyField =
  | Field<string, StringQuestionAdditionalProps>
  | Field<Date, DateQuestionAdditionalProps>;

export type AnyConfigType = { [key: string]: AnyField };

type FieldValue<T extends Field<any, any>> =
  T extends Field<infer V, any> ? V : never;

export type ObjectValue<C extends AnyConfigType> = {
  [K in keyof C]: FieldValue<C[K]>;
};

export type RawObjectValue<C extends AnyConfigType> = {
  [K in keyof C]: Maybe<FieldValue<C[K]>>;
};

export type ValidateFunction<Config extends AnyConfigType> = (
  input: ObjectValue<Config>,
) => { error: string; questionsToInvalidate: (keyof Config)[] } | null;
