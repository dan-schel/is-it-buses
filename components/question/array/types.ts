// TODO: In future, maybe we need indicesToInvalidate?
export type ArrayValidateFunction<Type> = (
  input: Type[],
) => { error: string } | null;
