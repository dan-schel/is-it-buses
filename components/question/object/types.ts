import { AnyField, Field } from "@/components/question/lib/field";
import { Maybe } from "@/shared/types/maybe";

export type AnyObjectConfig = { [key: string]: AnyField };

type FieldValue<T extends AnyField> =
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  T extends Field<infer V, any> ? V : never;

export type ObjectValue<Config extends AnyObjectConfig> = {
  [K in keyof Config]: FieldValue<Config[K]>;
};

export type RawObjectValue<Config extends AnyObjectConfig> = {
  [K in keyof Config]: Maybe<FieldValue<Config[K]>>;
};

export type ObjectValidateFunction<Config extends AnyObjectConfig> = (
  input: ObjectValue<Config>,
) => { error: string; questionsToInvalidate: (keyof Config)[] } | null;
