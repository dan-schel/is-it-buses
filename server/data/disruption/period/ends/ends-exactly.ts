import { z } from "zod";
import { DisplayStringOptions, EndsBase } from "./ends-base";
import { format, isSameYear } from "date-fns";

/** The disruption ends after an exact timestamp. */
export class EndsExactly extends EndsBase {
  constructor(readonly date: Date) {
    super();
  }

  static readonly bson = z
    .object({
      type: z.literal("exactly"),
      date: z.date(),
    })
    .transform((x) => new EndsExactly(x.date));

  toBson(): z.input<typeof EndsExactly.bson> {
    return {
      type: "exactly",
      date: this.date,
    };
  }

  getDisplayString(options: DisplayStringOptions): string {
    const formatCode = isSameYear(this.date, options.now)
      ? "h:mmaaa E do MMM"
      : "h:mmaaa E do MMM yyyy";
    return format(this.date, formatCode);
  }

  latestInterpretableDate(): Date | null {
    return this.date;
  }
}
