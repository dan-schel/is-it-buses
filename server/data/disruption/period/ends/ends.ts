import { z } from "zod";
import { EndsAfterLastService } from "@/server/data/disruption/period/ends/ends-after-last-service";
import { EndsApproximately } from "@/server/data/disruption/period/ends/ends-approximately";
import { EndsExactly } from "@/server/data/disruption/period/ends/ends-exactly";
import { EndsNever } from "@/server/data/disruption/period/ends/ends-never";
import { EndsWhenAlertEnds } from "@/server/data/disruption/period/ends/ends-when-alert-ends";

/**
 * Defines how/when the disruption ends.
 *
 * (All implementations of EndsBase.)
 */
export type Ends =
  | EndsAfterLastService
  | EndsApproximately
  | EndsExactly
  | EndsNever
  | EndsWhenAlertEnds;

export const endsBson = z.union([
  EndsAfterLastService.bson,
  EndsApproximately.bson,
  EndsExactly.bson,
  EndsNever.bson,
  EndsWhenAlertEnds.bson,
]);
