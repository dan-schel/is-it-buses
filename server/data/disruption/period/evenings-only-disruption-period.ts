import { z } from "zod";
import { DisruptionPeriodBase } from "./disruption-period-base";
import { Ends, endsBson } from "./ends/ends";

/** Disruption is active every evening from the start date to the end date. */
export class EveningsOnlyDisruptionPeriod extends DisruptionPeriodBase {
  constructor(
    public start: Date | null,
    public end: Ends,

    /** E.g. `18` for 6pm to last service each day. */
    public startHourEachDay: number,
  ) {
    super();
  }

  toDisplayString(): string {
    // TODO: I imagine these dates won't be accurate for every disruption.
    return `6pm to last service every evening, Sat 22 Feb to Sun 23 Feb`;
  }

  static readonly bson = z
    .object({
      type: z.literal("evenings-only"),
      start: z.date().nullable(),
      end: endsBson,
      startHourEachDay: z.number(),
    })
    .transform(
      (x) =>
        new EveningsOnlyDisruptionPeriod(x.start, x.end, x.startHourEachDay),
    );

  toBson(): z.input<typeof EveningsOnlyDisruptionPeriod.bson> {
    return {
      type: "evenings-only",
      start: this.start,
      end: this.end.toBson(),
      startHourEachDay: this.startHourEachDay,
    };
  }
}
