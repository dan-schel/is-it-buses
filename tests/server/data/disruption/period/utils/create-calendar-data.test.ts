import { EndsExactly } from "@/server/data/disruption/period/ends/ends-exactly";
import { EveningsOnlyDisruptionPeriod } from "@/server/data/disruption/period/evenings-only-disruption-period";
import { StandardDisruptionPeriod } from "@/server/data/disruption/period/standard-disruption-period";
import {
  daysToRenderOnCalendar,
  createCalendarData,
} from "@/server/data/disruption/period/utils/create-calendar-data";
import { CalendarCellData } from "@/shared/types/calendar-data";
import { describe, expect, it } from "vitest";

describe("#createCalendarData", () => {
  it("works", () => {
    const result = createCalendarData(
      [
        regular("2025-03-12", "2025-03-15"),
        evenings("2025-03-20", "2025-03-25"),
        regular("2025-03-24", "2025-03-26"),
        regular("2025-04-10", "2025-04-30"),
      ],
      new Date("2025-03-15T08:00:00+11:00"),
    );
    expect(result.cells.length).toBe(daysToRenderOnCalendar);
    expect(result.cells.map(toSnapshotString).join("\n")).toMatchSnapshot();
  });

  it("works with no periods given", () => {
    const result = createCalendarData(
      [],
      new Date("2025-03-15T08:00:00+11:00"),
    );
    expect(result.cells.length).toBe(daysToRenderOnCalendar);
    expect(result.cells.every((x) => x.mark === "no-disruption")).toBe(true);
  });

  it("handles timezones correctly", () => {
    const result1 = createCalendarData(
      [],
      new Date("2025-03-15T23:59:59+11:00"),
    );
    expect(result1.today).toEqual({ year: 2025, month: 3, day: 15 });
    expect(result1.cells.length).toBe(daysToRenderOnCalendar);
    expect(result1.cells[0].year).toBe(2025);
    expect(result1.cells[0].month).toBe(3);
    expect(result1.cells[0].day).toBe(15);

    const result2 = createCalendarData(
      [],
      new Date("2025-03-16T00:00:00+11:00"),
    );
    expect(result2.today).toEqual({ year: 2025, month: 3, day: 16 });
    expect(result2.cells.length).toBe(daysToRenderOnCalendar);
    expect(result2.cells[0].year).toBe(2025);
    expect(result2.cells[0].month).toBe(3);
    expect(result2.cells[0].day).toBe(16);
  });

  function regular(from: string, to: string) {
    return new StandardDisruptionPeriod(
      new Date(from + "T05:00:00+11:00"),
      new EndsExactly(new Date(to + "T22:00:00+11:00")),
    );
  }

  function evenings(from: string, to: string) {
    return new EveningsOnlyDisruptionPeriod(
      new Date(from + "T05:00:00+11:00"),
      new EndsExactly(new Date(to + "T22:00:00+11:00")),
      18,
    );
  }

  function toSnapshotString(mark: CalendarCellData): string {
    return `${pad(mark.year, 4)}-${pad(mark.month, 2)}-${pad(mark.day, 2)} ${mark.mark}`;
  }

  function pad(value: number, digits: number): string {
    return value.toFixed().padStart(digits, "0");
  }
});
