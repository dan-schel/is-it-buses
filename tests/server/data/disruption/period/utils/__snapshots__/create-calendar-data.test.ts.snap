// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`#createCalendarData > works 1`] = `
"2025-03-15 all-day
2025-03-16 no-disruption
2025-03-17 no-disruption
2025-03-18 no-disruption
2025-03-19 no-disruption
2025-03-20 evening-only
2025-03-21 evening-only
2025-03-22 evening-only
2025-03-23 evening-only
2025-03-24 all-day
2025-03-25 all-day
2025-03-26 all-day
2025-03-27 no-disruption
2025-03-28 no-disruption
2025-03-29 no-disruption
2025-03-30 no-disruption
2025-03-31 no-disruption
2025-04-01 no-disruption
2025-04-02 no-disruption
2025-04-03 no-disruption
2025-04-04 no-disruption
2025-04-05 no-disruption
2025-04-06 no-disruption
2025-04-07 no-disruption
2025-04-08 no-disruption
2025-04-09 no-disruption
2025-04-10 all-day
2025-04-11 all-day"
`;
